tripy/tripy/function_registry.py

```diff
@@ -197,14 +199,30 @@ def matches_type(name: str, annotation: type, arg: Any) -> bool:
         annotations = self._get_annotations()
 
         # Check if we have too many positional arguments. We can only do this if there isn't a variadic positional argument.
-        if not any(annotation.kind == inspect.Parameter.VAR_POSITIONAL for annotation in annotations.values()) and len(
-            args
-        ) > len(annotations):
+        annotation_items = list(annotations.items())
+        variadic_idx = None
+        for idx, (_, annotation) in enumerate(annotation_items):
+            # there can only be at most one variadic arg and it must come after all positional args and before keyword-only args
+            if annotation.kind == inspect.Parameter.VAR_POSITIONAL:
+                variadic_idx = idx
+                break
+
+        if variadic_idx is None and len(args) > len(annotations):
             return Result.err(
                 [f"Function expects {len(annotations)} parameters, but {len(args)} arguments were provided."],
             )
 
-        for (name, annotation), arg in zip(annotations.items(), args):
+        # If there is a variadic positional arg, we can copy the final annotation for the remaining args.
+        # Keyword-only args (only possible with a variadic arg) will appear in kwargs and don't need to be checked here.
+        if variadic_idx is not None:
+            positional_args_to_check = chain(
+                zip(annotation_items[:variadic_idx], args),
```

Comment:
We could actually just express it like this:
```suggestion
                zip(annotation_items, args),
```
`len(args)` will always be less than or equal to `len(annotation_items[:variadic_idx])` so we don't gain anything by slicing it in advance.